### ISSUE TITLE
Single Table Inheritance models with relationships

### ISSUE THREAD

**Issue by hector** [graphiti-api/graphiti_old_archived#56]
I have two models with single table inheritance that have a one to many association between them. I am getting errors when creating them using the API.
I suspect I may have errors on my resources. Could you tell me if the resources are correct? If they are I will post details about the error so we can dig on it.

Models:
```ruby
class Project < ApplicationRecord
  has_many :renderables
end
class DubbingProject < Project; end
class VoiceoverProject < Project; end

class Renderable < ApplicationRecord
  belongs_to :project
end
class Dubbing < Renderable; end
class Voiceover < Renderable; end
```

Resources:
```ruby
class ProjectResource < ApplicationResource
  self.polymorphic = %w(DubbingProjectResource VoiceoverProjectResource)
  has_many :renderables
end
class DubbingProjectResource < ProjectResource; end
class VoiceoverProjectResource < ProjectResource; end

class RenderableResource < ApplicationResource
  self.polymorphic = %w(DubbingResource VoiceoverResource)
  belongs_to :project
end
class DubbingResource < RenderableResource; end
class VoiceoverResource < RenderableResource; end
```

Schema:
```ruby
ActiveRecord::Schema.define do

  create_table "projects", force: :cascade do |t|
    t.string "type", null: false
  end

  create_table "renderables", force: :cascade do |t|
    t.bigint "project_id"
    t.string "type", null: false
  end

end
```

**Comment by richmolj** [comment]
This looks pretty good to me. I'm not sure the scenario has come up before so it's possible there's a bug - if you have a stacktrace or better yet a failing spec I can take more of a look.

**Comment by hector** [comment]
The error is produced when trying to save a Dubbing with a Project associated.

The stack trace is quite poor:
```
NoMethodError (undefined method `child_for_type' for #<#<Class:0x00007fcba8393cc8>:0x00007fcba972bf10>):
  
app/controllers/api_controller.rb:37:in `create'
```

This is the create method:
```ruby
  def create
    resource = graphiti_resource_class.build(params)

    if resource.save
      render jsonapi: resource, status: 201
    else
      render jsonapi_errors: resource
    end
  end
```

So I made some digging and the error is produced here:
https://github.com/graphiti-api/graphiti/blob/f07fb749d29c52f95dd522cb75c75a3ef4e3483a/lib/graphiti/util/relationship_payload.rb#L49

I solved the problem by commenting all the if block:
```ruby
if sideload.resource.type != type && sideload.polymorphic_parent?
  sideload = sideload.child_for_type(type)
end
```

I have the impression the purpose of this _if block_ is to change the `sideload` for polymorphic associations. The code checks if the types are different and if the parent is polymorphic but, sadly, this also hits my scenario. I think it is only a matter of only adding an extra condition to the _if_, any idea? It might be easier for you. If you have no clue I can dig deeper.

**Comment by hector** [comment]
I guess this would do:
```ruby
sideload.responds_to?(:child_for_type)
```
 But it doesn't seem very elegant

**Comment by richmolj** [comment]
@hector could you put your request here? I'm curious why the JSONAPI `type` would differ from the sideload `type`. Certainly this could happen if the jsonapi type was `dubbing_projects` in the payload for instance (totally valid just want to ensure this is the case).

You're correct the logic is to accommodate polymorphism, and `child_for_type` is present on polymorphic sideloads. The key thing here is that your relationship doesn't respond to that method but **we need it to**.

This is because you might sidepost further relationships. Imagine `DubbingProjectResource` could nest another association but `VoiceoverProjectResource` could not. Since the relationship only knows about `ProjectResource`, we need to replace it with the appropriate subclass so we know about further associations, different adapters, etc.

In other words, the same logic needs to fire for the same reasons, but the logic isn't currently implemented for these STI-type sideloads. I think we need to move the method from [here](https://github.com/graphiti-api/graphiti/blob/f07fb749d29c52f95dd522cb75c75a3ef4e3483a/lib/graphiti/sideload/polymorphic_belongs_to.rb#L75-L79) to [here](https://github.com/graphiti-api/graphiti/blob/f07fb749d29c52f95dd522cb75c75a3ef4e3483a/lib/graphiti/sideload.rb) so it's always available, and raise an error if the resource doesn't actually have polymorphic children. This should A) get your code working and B) raise a helpful error if the user is submitting an unsupported jsonapi type for the association.

Does this make sense?

**Comment by hector** [comment]
It does make sense.

Here is the complete payload. You'll see more attributes and relationships I didn't post before on my code for the sake of simplicity.
```json
{
  "data": {
    "type": "dubbings",
    "attributes": {
      "name": "Dubbing02",
      "original_subtitles": "eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBLQT09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--75f61f57c85c21cb0828f1250e9024738e550e56",
      "subtitles": "eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBLUT09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--9a0de6c7ee3b03869c156c273b4f92e217e58ed5",
      "video_subtitles": "eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBLZz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--7a8cf70749e7ca489120cf642c874db1b18009d4"
    },
    "relationships": {
      "project": {
        "data": {
          "type": "dubbing_projects",
          "id": "1",
          "method": "update"
        }
      },
      "language": {
        "data": {
          "type": "languages",
          "id": "1",
          "method": "update"
        }
      },
      "speakers": {
        "data": [
          {
            "type": "speakers",
            "temp-id": "temp-id-1",
            "method": "create"
          }
        ]
      }
    }
  },
  "included": [
    {
      "type": "dubbing_projects",
      "id": "1"
    },
    {
      "type": "languages",
      "id": "1"
    },
    {
      "type": "speakers",
      "temp-id": "temp-id-1",
      "attributes": {
        "name": "Speaker1"
      },
      "relationships": {
        "voice_model": {
          "data": {
            "type": "voice_models",
            "id": "1",
            "method": "update"
          }
        }
      }
    },
    {
      "type": "voice_models",
      "id": "1"
    }
  ]
}
```

**Comment by richmolj** [comment]
Sounds good - I think we're on the same page but not sure when I'll be able to get to it. Is this something where you'd be interested in submitting a PR? I think we mostly need a test in `spec/integration`, and the solution outlined in the prior comment.

**Comment by hector** [comment]
I will try if it is something I can handle

**Comment by hector** [comment]
Hi @richmolj, I have been checking the existing tests to figure out how to write this one but there is a lot of stuff going on and I would need a lot of time to decipher it. If this is something you could do I would be very grateful. I could prepare an example Rails application that fails, if this is something you think would help.

**Comment by richmolj** [comment]
A failing sample app would be super helpful! Not sure when I can get to it, but if I had that shouldnt be too hard

**Comment by hector** [comment]
Done!
https://github.com/hector/graphiti-issue-56

Find the instructions to run it in the readme.
